File: experiments/2026-8-26_operation-pruning-research/pruned_prng.c
// pruned_prng.c
// C port of pruned_prng.py for speed -- pure-Python generation proved too
// slow (~0.8 MB/s) to run PractRand tiers large enough to reliably catch
// broken candidates (see RESULTS.md "gate-size traps": 1MB and 8MB both let
// structurally broken candidates through that failed at 16-32MB). This port
// must be cross-validated bit-for-bit against pruned_prng.py's ALL_OPS
// (bitmask 0x3FFFF, hash_access=strided) before being trusted for anything,
// per the periodicity experiment's C-port precedent
// (experiments/2026-8-25_periodicity-heuristic-validation/cycle_measure.c).
//
// Op bitmask bit order (must match operations.py's canonical SCAN_ORDER --
// see that file / ablation_search.py for the authoritative list):
//   0-7:   TAP0..TAP7
//   8:     ROT_A   9: ROT_B   10: ROT_O   11: ROT_C
//   12:    SHL9    13: SHR18  14: SHL14   15: SHR13
//   16:    MULT_REDUCE
//   17:    HASH_SELFIDX
// hash_access: 0 = strided (original), 1 = sequential.
//
// Usage:
//   ./pruned_prng <seed> <iterations> <ops_bitmask_hex> <hash_access> [s9 s18 s14 s13] [--stream]
//       s9/s18/s14/s13 are the SHL9/SHR18/SHL14/SHR13 shift widths (default
//       9/18/14/13, the originals) -- only meaningful for whichever of those
//       flags are ON in the bitmask; added for the post-convergence "shift
//       repair" search (see RESULTS.md), which holds the op set fixed and
//       scans these widths for the point of maximum avalanche diffusion.
//   --stream streams raw little-endian uint32_t `c` values to stdout;
//   without it, prints the final `cons` value (for the self-check).

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#define ALL_OPS_MASK 0x3FFFFu

static inline uint32_t rot32(uint32_t n, uint32_t r) {
    r &= 31;
    return ((n << r) | (n >> (32 - r))) & 0xFFFFFFFFu;
}

static inline int has(uint32_t ops, int bit) {
    return (ops >> bit) & 1u;
}

// bit indices, matching operations.py's canonical order
enum {
    B_TAP0, B_TAP1, B_TAP2, B_TAP3, B_TAP4, B_TAP5, B_TAP6, B_TAP7,
    B_ROT_A, B_ROT_B, B_ROT_O, B_ROT_C,
    B_SHL9, B_SHR18, B_SHL14, B_SHR13,
    B_MULT_REDUCE, B_HASH_SELFIDX
};

static void ra_hash(const uint32_t *N, uint32_t *out8, uint32_t ops, int hash_access) {
    for (int i = 0; i < 8; ++i) {
        out8[i] = 0;
        if (has(ops, B_HASH_SELFIDX)) out8[i] ^= N[(uint8_t)N[i]];
        if (hash_access == 1) {
            int base = i * 32;
            for (int j = 0; j < 32; ++j) out8[i] ^= N[base + j];
        } else {
            for (int j = 0; j < 32; ++j) out8[i] ^= N[j * 8 + i];
        }
    }
}

static void init_state(uint32_t *L, uint32_t *M) {
    for (int i = 0; i < 256; ++i) {
        M[i] = (uint32_t)(i * 0x06a0dd9bUL + 0x06a0dd9bUL);
        L[i] = (uint32_t)(i * 0x9e3779b7UL + 0x9e3779b7UL);
    }
}

static inline uint32_t build_o(const uint32_t *M, uint32_t i, uint32_t ops) {
    uint32_t o = 0;
    for (uint32_t e = 0; e < 8; ++e) {
        if (has(ops, (int)e)) o ^= (M[(uint8_t)(i + e)] << e);
    }
    return o;
}

// One full 255-step permutation cycle. Writes each step's `c` to
// raw_stream (if non-NULL) as it's produced. Mutates L via the swap.
static void permutation_cycle(uint32_t cons, uint32_t it, uint32_t *M, uint32_t *L,
                               uint32_t ops, const uint32_t widths[4], FILE *raw_stream) {
    uint32_t a = cons, b = it, c = 0, d = 0;
    uint32_t s9 = widths[0], s18 = widths[1], s14 = widths[2], s13 = widths[3];

    for (uint32_t i = 255; i > 0; --i) {
        uint32_t o = build_o(M, i, ops);

        uint32_t raw_a = b ^ o;
        a = (has(ops, B_ROT_A) ? rot32(raw_a, d) : raw_a) ^ (cons + a);

        uint32_t raw_b = cons + a;
        b = (has(ops, B_ROT_B) ? rot32(raw_b, i) : raw_b) ^ (o + d);

        uint32_t rotated_o = has(ops, B_ROT_O) ? rot32(a ^ o, i) : (a ^ o);
        uint32_t left = has(ops, B_SHL9) ? (rotated_o << s9) : 0;
        uint32_t right = has(ops, B_SHR18) ? (b >> s18) : 0;
        if (has(ops, B_SHL9) && has(ops, B_SHR18)) o = left ^ right;
        else if (has(ops, B_SHL9)) o = left;
        else if (has(ops, B_SHR18)) o = right;
        else o = 0;

        uint32_t term_oc = has(ops, B_SHL14) ? ((o + c) << s14) : 0;
        uint32_t term_b13 = has(ops, B_SHR13) ? (b >> s13) : 0;
        uint32_t pre_rot_c = term_oc ^ term_b13 ^ a;
        c = has(ops, B_ROT_C) ? rot32(pre_rot_c, b) : pre_rot_c;

        if (raw_stream) fwrite(&c, sizeof(uint32_t), 1, raw_stream);

        d = has(ops, B_MULT_REDUCE) ? (uint32_t)(((uint64_t)c * (i + 1)) >> 32) : (c & 0xFFu);

        uint32_t tmp = L[i];
        L[i] = L[d];
        L[d] = tmp;
    }
}

static uint32_t reseed(uint32_t *M, const uint32_t *L, uint32_t ops, int hash_access) {
    for (int i = 0; i < 256; ++i) M[i] ^= L[i];

    uint32_t tmp8[8];
    ra_hash(M, tmp8, ops, hash_access);

    uint32_t new_cons = 0;
    for (int e = 0; e < 8; ++e) new_cons ^= (tmp8[e] << e);
    return new_cons;
}

static uint32_t run(uint32_t seed, size_t iterations, uint32_t ops, int hash_access,
                     const uint32_t widths[4], FILE *raw_stream) {
    static uint32_t L[256], M[256];
    init_state(L, M);
    uint32_t cons = seed;
    for (size_t it = 0; it < iterations; ++it) {
        permutation_cycle(cons, (uint32_t)it, M, L, ops, widths, raw_stream);
        cons = reseed(M, L, ops, hash_access);
    }
    return cons;
}

int main(int argc, char **argv) {
    if (argc < 5) {
        fprintf(stderr, "usage: %s <seed> <iterations> <ops_bitmask_hex> <hash_access:0|1> "
                         "[s9 s18 s14 s13] [--stream]\n", argv[0]);
        return 1;
    }
    uint32_t seed = (uint32_t)strtoul(argv[1], NULL, 0);
    size_t iterations = (size_t)strtoull(argv[2], NULL, 0);
    uint32_t ops = (uint32_t)strtoul(argv[3], NULL, 16);
    int hash_access = atoi(argv[4]);

    uint32_t widths[4] = {9, 18, 14, 13};
    int next = 5;
    if (argc >= 9 && strcmp(argv[5], "--stream") != 0) {
        widths[0] = (uint32_t)strtoul(argv[5], NULL, 0);
        widths[1] = (uint32_t)strtoul(argv[6], NULL, 0);
        widths[2] = (uint32_t)strtoul(argv[7], NULL, 0);
        widths[3] = (uint32_t)strtoul(argv[8], NULL, 0);
        next = 9;
    }
    int stream_mode = (argc > next && strcmp(argv[next], "--stream") == 0);

    uint32_t final_cons = run(seed, iterations, ops, hash_access, widths, stream_mode ? stdout : NULL);

    if (!stream_mode) {
        printf("%u\n", final_cons);
    }
    return 0;
}

File: experiments/2026-8-26_operation-pruning-research/pruned_prng.py
"""Operation-flag-parameterized reimplementation of ra_prng2's permutation
cycle + reseed, at full scale (n=256, w=32).

Ported from src/ra_prng2/python/ra_prng2.py (the paper-exact reference,
untouched) and experiments/others/avalanche_effect_analysis.py's
zepfold_capture() (the existing full-scale c-sequence-capture pattern).

This file is the sandbox: it is allowed to change output. It must NOT be
confused with the reference -- run this module directly (`python
pruned_prng.py`) to self-check that ALL_OPS reproduces the reference
bit-for-bit before trusting any ablation result.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from operations import (
    ALL_FLAGS, HASH_DEPTH_FIXED, HASH_ACCESS_DEFAULT, HASH_ACCESS_STRIDED,
    HASH_ACCESS_SEQUENTIAL, MULT_REDUCE, HASH_SELFIDX,
    ROT_A, ROT_B, ROT_O, ROT_C, SHL9, SHR18, SHL14, SHR13,
)

MASK = 0xFFFFFFFF


def rot32(n: int, r: int) -> int:
    r &= 31
    n &= MASK
    return ((n << r) | (n >> (32 - r))) & MASK


DEFAULT_SHIFT_WIDTHS = (9, 18, 14, 13)  # (SHL9, SHR18, SHL14, SHR13), original values


@dataclass(frozen=True)
class Candidate:
    """One ablation candidate: which flags are active, and the hash access
    pattern. HASH_DEPTH is deliberately not a field here -- it is fixed at
    32 everywhere (see operations.py's HASH_DEPTH_FIXED docstring for why).

    shift_widths: (s9, s18, s14, s13) -- the actual shift amounts used by
    whichever of SHL9/SHR18/SHL14/SHR13 are active in `ops`. Defaults to the
    original algorithm's constants. Only meaningful for flags that are ON;
    an OFF flag's width value is irrelevant (that XOR term is dropped
    entirely regardless of width, per operations.py's "off" semantics).
    Used for the post-convergence "shift repair" search: with the winning
    candidate's op set held fixed, scan these widths for the point of
    maximum avalanche diffusion ("chaos"), independent of which ops are on.
    """
    ops: frozenset[str] = field(default_factory=lambda: frozenset(ALL_FLAGS))
    hash_access: str = HASH_ACCESS_DEFAULT
    shift_widths: tuple[int, int, int, int] = DEFAULT_SHIFT_WIDTHS

    def has(self, flag: str) -> bool:
        return flag in self.ops


def ra_hash(N: list[int], cand: Candidate) -> list[int]:
    out = [0] * 8
    for i in range(8):
        if cand.has(HASH_SELFIDX):
            out[i] ^= N[N[i] & 0xFF]
        if cand.hash_access == HASH_ACCESS_SEQUENTIAL:
            base = i * HASH_DEPTH_FIXED
            for j in range(HASH_DEPTH_FIXED):
                out[i] ^= N[base + j]
        else:  # HASH_ACCESS_STRIDED (original)
            for j in range(HASH_DEPTH_FIXED):
                out[i] ^= N[j * 8 + i]
    return out


def init_state():
    M = [0] * 256
    L = [0] * 256
    for i in range(256):
        M[i] = (i * 0x06a0dd9b + 0x06a0dd9b) & MASK
        L[i] = (i * 0x9e3779b7 + 0x9e3779b7) & MASK
    return L, M


def _build_o(M: list[int], i: int, cand: Candidate) -> int:
    o = 0
    for e in range(8):
        if cand.has(f"TAP{e}"):
            o ^= (M[(i + e) & 0xFF] << e) & MASK
    return o


def permutation_cycle(cons: int, it: int, M: list[int], L: list[int], cand: Candidate):
    """One full 255-step permutation cycle. Yields each step's `c` (the real
    per-step RNG output) and mutates L in place via the swap, mirroring
    ra_permutation_cycle in src/ra_prng2/c/ra_prng2.c exactly when
    cand == ALL_OPS.
    """
    a = cons
    b = it
    c = 0
    d = 0

    for i in range(255, 0, -1):
        o = _build_o(M, i, cand)

        raw_a = (b ^ o) & MASK
        a = (rot32(raw_a, d) if cand.has(ROT_A) else raw_a) ^ ((cons + a) & MASK)
        a &= MASK

        raw_b = (cons + a) & MASK
        b = (rot32(raw_b, i) if cand.has(ROT_B) else raw_b) ^ ((o + d) & MASK)
        b &= MASK

        s9, s18, s14, s13 = cand.shift_widths
        rotated_o = rot32(a ^ o, i) if cand.has(ROT_O) else (a ^ o)
        left = ((rotated_o << s9) & MASK) if cand.has(SHL9) else 0
        right = (b >> s18) if cand.has(SHR18) else 0
        if cand.has(SHL9) and cand.has(SHR18):
            o = left ^ right
        elif cand.has(SHL9):
            o = left
        elif cand.has(SHR18):
            o = right
        else:
            o = 0

        term_oc = ((o + c) << s14) & MASK if cand.has(SHL14) else 0
        term_b13 = (b >> s13) if cand.has(SHR13) else 0
        pre_rot_c = term_oc ^ term_b13 ^ a
        c = (rot32(pre_rot_c, b) if cand.has(ROT_C) else pre_rot_c) & MASK

        yield c

        if cand.has(MULT_REDUCE):
            d = (c * (i + 1)) >> 32
        else:
            d = c & 0xFF

        L[i], L[d] = L[d], L[i]


def reseed(M: list[int], L: list[int], cand: Candidate) -> int:
    for i in range(256):
        M[i] ^= L[i]

    tmp8 = ra_hash(M, cand)
    new_cons = 0
    for e in range(8):
        new_cons ^= (tmp8[e] << e) & MASK
    return new_cons


def stream(seed: int, iterations: int, cand: Candidate):
    """Yield every `c` value across `iterations` full permutation cycles,
    reseeding between each -- the same shape as ra_core's outer loop, minus
    the mid-cycle truncation-count logic (which only matters for the last
    partial cycle of a specific `rng` total, irrelevant for quality testing
    a long stream).
    """
    L, M = init_state()
    cons = seed
    for it in range(iterations):
        yield from permutation_cycle(cons, it, M, L, cand)
        cons = reseed(M, L, cand)


def final_cons(seed: int, iterations: int, cand: Candidate) -> int:
    cons = seed
    L, M = init_state()
    for it in range(iterations):
        for _ in permutation_cycle(cons, it, M, L, cand):
            pass
        cons = reseed(M, L, cand)
    return cons


def _self_check():
    """ALL_OPS must reproduce src/ra_prng2/python/ra_prng2.py bit-for-bit."""
    import sys
    sys.path.insert(0, "../../src/ra_prng2/python")
    from ra_prng2 import ra_core as reference_ra_core

    all_ops = Candidate()
    for seed, iterations in [(1, 1), (1, 2), (42, 3), (0xDEADBEEF, 1)]:
        expected = reference_ra_core(seed, iterations)
        got = final_cons(seed, iterations, all_ops)
        status = "OK" if expected == got else "MISMATCH"
        print(f"seed={seed} iterations={iterations}: expected={expected} got={got} [{status}]")
        assert expected == got, "pruned_prng.py with ALL_OPS must match the paper-exact reference"
    print("Self-check passed: ALL_OPS is bit-identical to src/ra_prng2/python/ra_prng2.py")


if __name__ == "__main__":
    _self_check()
